apps/frontend/src/app/PageArtifact/DupModal.tsx

```diff
@@ -0,0 +1,90 @@
+import {
+  Alert,
+  Box,
+  CardContent,
+  Divider,
+  Stack,
+  Typography,
+} from '@mui/material'
+import { useContext, useEffect, useMemo } from 'react'
+import { useTranslation } from 'react-i18next'
+import CardDark from '../Components/Card/CardDark'
+import CardLight from '../Components/Card/CardLight'
+import CloseButton from '../Components/CloseButton'
+import ModalWrapper from '../Components/ModalWrapper'
+import { DatabaseContext } from '../Database/Database'
+import useForceUpdate from '../ReactHooks/useForceUpdate'
+import ArtifactCard from './ArtifactCard'
+import DifferenceIcon from '@mui/icons-material/Difference'
+export default function DupModal({ show, onHide }) {
+  const { t } = useTranslation('artifact')
+  return (
+    <ModalWrapper open={show} onClose={onHide}>
+      <CardLight>
+        <CardContent sx={{ display: 'flex', alignItems: 'center' }}>
+          <Typography variant="h6" flexGrow={1}>
+            <DifferenceIcon sx={{ verticalAlign: 'text-top', mr: 1 }} />
+            {t`showDup`}
+          </Typography>
+          <CloseButton onClick={onHide} />
+        </CardContent>
+        <Divider />
+        <CardContent>
+          <DupContent />
+        </CardContent>
+      </CardLight>
+    </ModalWrapper>
+  )
+}
+function DupContent() {
+  const { t } = useTranslation('artifact')
+  const { database } = useContext(DatabaseContext)
+  const [dbDirty, setDBDirty] = useForceUpdate()
+  useEffect(() => database.arts.followAny(setDBDirty), [setDBDirty, database])
+
+  const dupList = useMemo(() => {
+    const dups = dbDirty && ([] as Array<Array<string>>)
+    let artKeys = database.arts.keys
+
+    while (artKeys.length !== 0) {
```

Comment:
Why this over `for (const artKey of artKeys)` or `artKeys.forEach()`?

Reply:
How? Also can you add a comment then

Reply:
`artKeys` is a array of ids that hasnt gotten any dups. when dups are found, all occurence gets removed from the list.

Reply:
Maybe I'm missing something, I don't see how that is happening with [Array.shift()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/shift)

Reply:
`Array.shift` removes the 1st element of the array and returns it. 
If dups are detected from that element, those are removed as well.

Reply:
oh i totally missed line 63 :)